In [3]:
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

products_df = pd.read_csv("data/olist_products_dataset.csv")

products_df.head()


product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [4]:
## Check the number of unique rows based on product id

In [5]:
#Total unique rows based on the product id
len(products_df["product_id"].unique())


32951

In [6]:
## Drop rows that have similar values

In [7]:
#Check for duplicates rows
len(products_df.drop_duplicates())


32951

## Information on the dataframe - data type  for each column

In [8]:
#information about the dataframe 

products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


## Generate descriptive statistics for a DataFrame named products_df. This function provides summary statistics such as count, mean, standard deviation, minimum, maximum, and quartiles for numerical columns in the DataFrame.

In [9]:
#describe the dataframe
products_df.describe()


product_name_lenght  product_description_lenght  product_photos_qty  \
count         32341.000000                32341.000000        32341.000000   
mean             48.476949                  771.495285            2.188986   
std              10.245741                  635.115225            1.736766   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32949.000000       32949.000000       32949.000000   
mean        2276.472488          30.815078          16.937661   
std         4282.038731          16.914458          13.637554   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1900.000000          38.000000          21.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count      32949.000000  
mean          23.196728  
std           12.079047  
min            6.000000  
25%           15.000000  
50%           20.000000  
75%           30.000000  
max          118.000000

## Display the number of rows and columns

In [10]:
# Get the count of total rows and columns
products_df.shape

(32951, 9)

## Display the sum of null values in the dataframe

In [11]:
# Calculate the sum of empty cells in each column
print("\nEmpty cells count in DataFrame:")
print(products_df.isnull().sum())



Empty cells count in DataFrame:
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64


## Display the NaN values in the product_category_name

In [12]:
#display null values product_category_name
products_df[products_df["product_category_name"].isnull()].head()


product_id product_category_name  \
105  a41e356c76fab66334f36de622ecbd3a                   NaN   
128  d8dee61c2034d6d075997acef1870e9b                   NaN   
145  56139431d72cd51f19eb9f7dae4d1617                   NaN   
154  46b48281eb6d663ced748f324108c733                   NaN   
197  5fb61f482620cb672f5e586bb132eae9                   NaN   

     product_name_lenght  product_description_lenght  product_photos_qty  \
105                  NaN                         NaN                 NaN   
128                  NaN                         NaN                 NaN   
145                  NaN                         NaN                 NaN   
154                  NaN                         NaN                 NaN   
197                  NaN                         NaN                 NaN   

     product_weight_g  product_length_cm  product_height_cm  product_width_cm  
105             650.0               17.0               14.0              12.0  
128             300.0               16.0                7.0              20.0  
145             200.0               20.0               20.0              20.0  
154           18500.0               41.0               30.0              41.0  
197             300.0               35.0                7.0              12.0

## Reading of products_category_name_translation file

In [13]:
#Read from product_category_name_translation

product_category_translated_df= pd.read_csv("data/product_category_name_translation.csv")

product_category_translated_df.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

## Extract the translated word from product_category_translated and insert into a new column into products_df

In [14]:
#Extract the  matching translation word from nametranslation_df to into a new column into products_df

products_df['product_category_name_translated'] = products_df['product_category_name'].map(product_category_translated_df.set_index('product_category_name')['product_category_name_english'])

# Move the translated column next to product_category_name for easy referencing
# Get the current index of the column
current_index = products_df.columns.get_loc("product_category_name_translated")

# Remove the column from its current position and insert it at index 2
column = products_df.pop("product_category_name_translated")
products_df.insert(2, "product_category_name_translated", column)
products_df.head()


product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

  product_category_name_translated  product_name_lenght  \
0                        perfumery                 40.0   
1                              art                 44.0   
2                   sports_leisure                 46.0   
3                             baby                 27.0   
4                       housewares                 37.0   

   product_description_lenght  product_photos_qty  product_weight_g  \
0                       287.0                 1.0             225.0   
1                       276.0                 1.0            1000.0   
2                       250.0                 1.0             154.0   
3                       261.0                 1.0             371.0   
4                       402.0                 4.0             625.0   

   product_length_cm  product_height_cm  product_width_cm  
0               16.0               10.0              14.0  
1               30.0               18.0              20.0  
2               18.0                9.0              15.0  
3               26.0                4.0              26.0  
4               20.0               17.0              13.0

## Replace nan values in product category name to unclassifed and replace all other nan values with nan

In [15]:



# # Replace NaN (null) values in the specified column with 'unclassified'
products_df['product_category_name'].fillna('unclassified', inplace=True)


# #Replace Nan (null) values wth nan in the specificed column with'NaN

products_df.fillna('NaN', inplace=True)



C:\Users\User\AppData\Local\Temp\ipykernel_8736\917828273.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  products_df['product_category_name'].fillna('unclassified', inplace=True)
C:\Users\User\AppData\Local\Temp\ipykernel_8736\917828273.py:7: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'NaN' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  products_df.fillna('NaN', inplace=True)


## Connecton to postgresql

In [16]:
# connection to postgresql 




conn =None
cur=None


try:
    conn = psycopg2.connect(
    host='localhost',
    dbname='Database(new)',
    user='postgres',
    password='admin',
    port='5432',
    )
    cur=conn.cursor()

    create_script_review ='''CREATE TABLE IF NOT EXISTS olist_products_dateset (
        product_id varchar(200) primary key,
        product_category_name	varchar(200),
        translated_word	varchar(200),
        product_name_lenght	 DECIMAL(10,1),
        product_description_lenght  DECIMAL(10,1),
        product_photos_qty	 DECIMAL(10,1),
        product_weight_g	 DECIMAL(10,1),
        product_length_cm	 DECIMAL(10,1),
        product_height_cm	 DECIMAL(10,1),
        product_width_cm     DECIMAL(10,1)
        FOREIGN KEY (olist_order_items_dataset) REFERENCES other_table(product_id)
 )'''


    cur.execute(create_script_review)

    conn.commit()
    for index, row in products_df.iterrows():
       
        insert_script_host = '''INSERT INTO olist_products  (  
        product_id ,
        product_category_name	,
        translated_word,
        product_name_lenght,
        product_description_lenght ,
        product_photos_qty,
        product_weight_g	,
        product_length_cm,
        product_height_cm,
        product_width_cm) VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s) '''
        values =(row.iloc[0],row.iloc[1],row.iloc[2],row.iloc[3],row.iloc[4],row.iloc[5],row.iloc[6],row.iloc[7],row.iloc[8],row.iloc[9])
        cur.execute(insert_script_host,values)
        conn.commit()


except Exception as error:      
    print(error)


 finally:

      if cur is not None:

           cur.close()
#           conn.close()


Remarks:

- Product ID is unique no duplicates found
-No duplicates rows has been found
-Total unique records: 32951

- Data types of the products_df dataframe

- 0   product_id                  32951 non-null  object 
- 1   product_category_name       32341 non-null  object 
-  2   product_name_lenght         32341 non-null  float64
- 3   product_description_lenght  32341 non-null  float64
- 4   product_photos_qty          32341 non-null  float64
- 5   product_weight_g            32949 non-null  float64
-6   product_length_cm           32949 non-null  float64
- 7   product_height_cm           32949 non-null  float64
- 8   product_width_cm            32949 non-null  float64

No product category name given means not set by the user (replace with unclassifed )
- No duplicate rows found ( each product is unique)
- Check for empty cell in each column (empty cells for category and measurement might not be set by user)
-Extract the  matching translation word from nametranslation_df to into a new column into products_df
-Based on min and max values in the measurements, whole number part is set to max 5 digits and decimal part is set one decimal point (will be done in the upload into postgre part)

Tranformation:
- Extract the  matching translation word from nametranslation_df to into a new column into products_df
- Move the translated column next to product_category_name for easy referencing
- Replace NaN (null) values in the specified column  product category name with 'unclassified'
- Replace Nan (null) values wth nan in the specificed column with'NaN






